crates/commitlog/src/commitlog.rs

```diff
@@ -279,6 +251,104 @@ impl<R: Repo, T> Drop for Generic<R, T> {
     }
 }
 
+pub fn commits_from<R: Repo>(repo: R, max_log_format_version: u8, offset: u64) -> io::Result<Commits<R>> {
+    let mut offsets = repo.existing_offsets()?;
+    if let Some(pos) = offsets.iter().rposition(|&off| off <= offset) {
+        offsets = offsets.split_off(pos);
+    }
+    let last_offset = offsets.first().cloned().unwrap_or(offset);
+    let segments = Segments {
+        offs: offsets.into_iter(),
+        repo,
+        max_log_format_version,
+    };
+    Ok(Commits {
+        inner: None,
+        segments,
+        last_offset,
+        last_error: None,
+    })
+}
+
+pub fn transactions_from<'a, R, D, T>(
```

Comment:
Is it worth having these transactions_from variants here? Maybe I'm not aware of existing code to persist the tables schema state, or how to skip entire tx ranges using these functions, but it seems that unless the caller can persist their decoder state, they always have to replay from 0. In the case of code under dev, where a full replay is needed on code changes, going through the full snapshot on every iteration is too slow, and --release isn't friendly inside the debugger.

For analytics, the custom decoder calls `decode_record_fn` and only replays records for the system tables to create the schemas needed to decode rows, and processes everything else on the fly from within the Visitor.

For information, I'm only using `commits`. This is because, given the large snapshot at the beginning of the log, it's starting at offset 0 to load the schema information, then skips to offset 163 to load the relevant table's snapshot, and then skips all the way to the first post-snapshot commit. This saves most of the startup time without having to persist the state of the schemas, ie many millions of records don't have to be decoded at all before getting to the interesting data.

Reply:
Ah nevermind most of that, just realized Decoder.decode_record is meant to call decode_record_fn, and the tx could be filtered there instead.

Reply:
`_from` is useful for forensics, and for replication.